models/classes/runner/plugins/TestPlugin.php

```diff
@@ -0,0 +1,229 @@
+     */
+    public function __construct ($id, $module, $category, $data = [] )
+    {
+
+        self::validateRequiredData($id, $module, $category);
+
+        $this->id          = (string)  $id;
+        $this->module      = (string)  $module;
+        $this->category    = (string)  $category;
+
+        if(isset($data['bundle'])) {
+            $this->bundle  = (string)  $data['bundle'];
+        }
+        if(isset($data['position'])) {
+            $this->position  = $data['position'];
+        }
+        if(isset($data['description'])) {
+            $this->description  = (string) $data['description'];
+        }
+        if(isset($data['name'])) {
+            $this->name  = (string) $data['name'];
+        }
+        if(isset($data['active'])) {
+            $this->active = (boolean) $data['active'];
+        }
+        if(isset($data['tags'])) {
+            $this->tags = (array) $data['tags'];
+        }
+    }
+
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    public function getModule()
+    {
+        return $this->module;
+    }
+
+    public function getBundle()
+    {
+        return $this->bundle;
+    }
+
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    public function getDescription()
+    {
+        return $this->description;
+    }
+
+    public function getCategory()
+    {
+        return $this->category;
+    }
+
+    public function getPosition()
+    {
+        return $this->position;
+    }
+
+    public function isActive()
+    {
+        return $this->active;
+    }
+
+    public function setActive($active)
+    {
+        $this->active = (boolean) $active;
+    }
+
+    public function getTags()
+    {
+        return $this->tags;
+    }
+
+    public function hasTag($tag)
+    {
+        return in_array($this->tags, $tag);
+    }
+
+    /**
+     * @see JsonSerializable::jsonSerialize
+     */
+    public function jsonSerialize()
+    {
+        return [
+            'id'          => $this->id,
+            'module'      => $this->module,
+            'bundle'      => $this->bundle,
+            'position'    => $this->position,
+            'name'        => $this->name,
+            'description' => $this->description,
+            'category'    => $this->category,
+            'active'      => $this->active,
+            'tags'        => $this->tags
+        ];
+    }
+
+    /**
+     * Create a test plugin from an assoc array
+     * @param array $data
+     * @return TestPlugin the new instance
+     * @throws common_exception_InconsistentData
+     */
+    public static function fromArray( array $data )
+    {
+
+        if( !isset($data['id']) || !isset($data['module']) || !isset($data['category']) ) {
+            throw new common_exception_InconsistentData('The plugin requires an id, a module and a category');
+        }
+        if(self::validateRequiredData($data['id'], $data['module'], $data['category'])){
```

Comment:
you're right. So you are doing the test twice and in all case you will throw an (uncatched) exception. You could directly try the ```new self``` without ```self::validateRequiredData``` before.